File: server/src/main/scala/producer/message_generator/data_generators/bytes/BytesGenerator.scala
package producer.message_generator.data_generators.bytes

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import org.graalvm.polyglot.Context

case class BytesGenerator(
    generator: RandomBytesGenerator | BytesFromBase64Generator | BytesFromHexGenerator
):
    def generate(polyglotContext: Context): Array[Byte] = generator match
        case v: RandomBytesGenerator =>
            v.generate
        case v: BytesFromBase64Generator =>
            v.generate(polyglotContext)
        case v: BytesFromHexGenerator =>
            v.generate(polyglotContext)

object BytesGenerator:
    def fromPb(v: pb.BytesGenerator): BytesGenerator =
        val generator = v.generator match
            case pb.BytesGenerator.Generator.GeneratorRandomBytes(v) =>
                RandomBytesGenerator.fromPb(v)
            case pb.BytesGenerator.Generator.GeneratorFromBase64(v) =>
                BytesFromBase64Generator.fromPb(v)
            case pb.BytesGenerator.Generator.GeneratorFromHex(v) =>
                BytesFromHexGenerator.fromPb(v)

        BytesGenerator(
            generator = generator
        )

    def toPb(v: BytesGenerator): pb.BytesGenerator =
        val generator = v.generator match
            case v: RandomBytesGenerator =>
                pb.BytesGenerator.Generator.GeneratorRandomBytes(RandomBytesGenerator.toPb(v))
            case v: BytesFromBase64Generator =>
                pb.BytesGenerator.Generator.GeneratorFromBase64(BytesFromBase64Generator.toPb(v))
            case v: BytesFromHexGenerator =>
                pb.BytesGenerator.Generator.GeneratorFromHex(BytesFromHexGenerator.toPb(v))

        pb.BytesGenerator(
            generator = generator
        )

File: ui/components/ui/ProducerSession/message-generator/data-generators/bytes/bytes-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { BytesFromBase64Generator, bytesFromBase64GeneratorFromPb } from './bytes-from-base64-generator';
import { BytesFromHexGenerator, bytesFromHexGeneratorFromPb } from './bytes-from-hex-generator';
import { RandomBytesGenerator, randomBytesGeneratorFromPb } from './random-bytes-generator';

export type BytesGenerator = {
  type: 'bytes-generator',
  generator: RandomBytesGenerator | BytesFromBase64Generator | BytesFromHexGenerator
};

export function bytesGeneratorFromPb(v: pb.BytesGenerator): BytesGenerator {
  switch (v.getGeneratorCase()) {
    case pb.BytesGenerator.GeneratorCase.GENERATOR_RANDOM_BYTES:
      return {
        type: 'bytes-generator',
        generator: randomBytesGeneratorFromPb(v.getGeneratorRandomBytes()!)
      };
    case pb.BytesGenerator.GeneratorCase.GENERATOR_FROM_BASE64:
      return {
        type: 'bytes-generator',
        generator: bytesFromBase64GeneratorFromPb(v.getGeneratorFromBase64()!)
      };
    case pb.BytesGenerator.GeneratorCase.GENERATOR_FROM_HEX:
      return {
        type: 'bytes-generator',
        generator: bytesFromHexGeneratorFromPb(v.getGeneratorFromHex()!)
      };
    default:
      throw new Error('Unknown bytes generator case');
  }
}
